tty/pc-tty/ttypc_vga.c

```diff
@@ -1,119 +1,258 @@
+	}
+	else {
+		vt->scrbsz += n;
+	}
+	memcpy(vt->scrb + (vt->scrbsz - n) * vt->cols, vt->vram + vt->top * vt->cols, n * vt->cols * CHR_VGA);
+
+	/* Roll up */
+	memmove(vt->vram + vt->top * vt->cols, vt->vram + (vt->top + n) * vt->cols, (vt->bottom - vt->top - n + 1) * vt->cols * CHR_VGA);
+	memsetw(vt->vram + (vt->bottom - n + 1) * vt->cols, vt->attr | ' ', n * vt->cols);
+}
 
 
-enum { crtcCursorH = 0xe, crtcCursorL = 0xf };
+void _ttypc_vga_rolldown(ttypc_vt_t *vt, unsigned int n)
+{
+	unsigned int k;
+
+	/* Roll down */
+	if (vt->bottom > vt->crow) {
+		k = min(n, vt->bottom - vt->crow);
+		memmove(vt->vram + (vt->top + k) * vt->cols, vt->vram + vt->top * vt->cols, (vt->bottom - vt->top - k + 1) * vt->cols * CHR_VGA);
+		memsetw(vt->vram + vt->top * vt->cols, vt->attr | ' ', k * vt->cols);
+
+		if (n == k)
+			return;
+		n -= k;
+	}
+
+	/* Update scrollback buffer */
+	if (vt->scrbsz + n > vt->rows * SCRB_PAGES) {
+		/* Make space for n new lines */
+		memmove(vt->scrb, vt->scrb + (vt->rows * SCRB_PAGES + 1 - vt->scrbsz) * vt->cols, (2 * vt->scrbsz - vt->rows * SCRB_PAGES - 1) * vt->cols * CHR_VGA);
+		vt->scrbsz = vt->rows * SCRB_PAGES;
+	}
+	else {
+		vt->scrbsz += n;
+	}
+	memsetw(vt->scrb + (vt->scrbsz - n) * vt->cols, vt->attr | ' ', n * vt->cols);
+}
+
+
+void _ttypc_vga_scrollup(ttypc_vt_t *vt, unsigned int n)
+{
+	/* Clip n */
+	if (n > vt->scrbsz - vt->scrbpos) {
+		if (!(vt->scrbsz - vt->scrbpos))
+			return;
+		n = vt->scrbsz - vt->scrbpos;
+	}
+
+	if (!n)
+		return;
+
+	/* Save scroll origin */
+	if (!vt->scrbpos)
+		memcpy(vt->scro, vt->vram, vt->rows * vt->cols * CHR_VGA);
+
+	/* Copy scrollback */
+	vt->scrbpos += n;
+	memcpy(vt->vram, vt->scrb + (vt->scrbsz - vt->scrbpos) * vt->cols, min(vt->scrbpos, vt->rows) * vt->cols * CHR_VGA);
+
+	/* Copy scroll origin */
+	if (vt->scrbpos < vt->rows)
+		memcpy(vt->vram + vt->scrbpos * vt->cols, vt->scro, (vt->rows - vt->scrbpos) * vt->cols * CHR_VGA);
+
+	/* Update cursor */
+	vt->cpos += n * vt->cols;
+	vt->crow += n;
+
+	if (vt == vt->ttypc->vt)
+		_ttypc_vga_setcursor(vt);
+}
 
 
-void _ttypc_vga_cursor(ttypc_virt_t *virt)
+void _ttypc_vga_scrolldown(ttypc_vt_t *vt, unsigned int n)
 {
-	ttypc_t *ttypc = virt->ttypc;
+	/* Clip n */
+	if (n > vt->scrbpos) {
+		if (!vt->scrbpos)
+			return;
+		n = vt->scrbpos;
+	}
 
-	outb(ttypc->out_crtc, crtcCursorH);
-	outb(ttypc->out_crtc + 1, virt->cur_offset >> 8);
-	
-	outb(ttypc->out_crtc, crtcCursorL);
-	outb(ttypc->out_crtc + 1, virt->cur_offset & 0xff);
+	if (!n)
+		return;
+
+	/* Copy scrollback */
+	vt->scrbpos -= n;
+	memcpy(vt->vram, vt->scrb + (vt->scrbsz - vt->scrbpos) * vt->cols, min(vt->scrbpos, vt->rows) * vt->cols * CHR_VGA);
+
+	/* Copy scroll origin */
+	if (vt->scrbpos < vt->rows)
+		memcpy(vt->vram + vt->scrbpos * vt->cols, vt->scro, (vt->rows - vt->scrbpos) * vt->cols * CHR_VGA);
+
+	/* Update cursor */
+	vt->cpos -= n * vt->cols;
+	vt->crow -= n;
+
+	if (vt == vt->ttypc->vt)
+		_ttypc_vga_setcursor(vt);
 }
```

Comment:
Almost identical code. Maybe `n` should be signed and one function to rule them all?

Reply:
Fixed.